napari/_qt/qt_viewer.py

```diff
@@ -1034,6 +1062,27 @@ def toggle_console_visibility(self, event=None):
             self.viewerButtons.consoleButton
         )
 
+    def on_fps_update(self, event):
+        if not self._camera_monitor.camera_moving:
+            return
+        fps = event.fps
+        if fps < self.viewer.target_fps[0]:
+            logging.info('decrease render quality')
+            for layer in self.viewer.layers:
+                layer.change_render_quality(RenderQualityChange.DECREASE)
+        if fps > self.viewer.target_fps[1]:
+            logging.info('increase render quality')
+            for layer in self.viewer.layers:
+                layer.change_render_quality(RenderQualityChange.INCREASE)
+
```

Comment:
This smells!

Reply:
@brisvag if a layer attribute says "adaptive", how do you propose the layer adapts? I think these things should live on the layer because many adaptive things can be done on the layer — subsample, switch to a lower res level in multiscales, switch to a decimated mesh in surfaces... The first implementation of this *happens* to be on the VisPy level but in general it could be at any level...

Reply:
(ie, I actually don't think it smells, I think doing it all in VisPy-land smells — for some quality changes it may be too late.)

Reply:
I think we're talking about different things :sweat_smile: I'm not saying to do it in vispy land, just to do it on a "viewer" level: why does every layer have the same value as every other layer?? That's just a global value at that point...

By `Adaptive` I mean this behaviour implemented here (fps based). `Performance` means "always render at low quality" and `Quality` means "always render at high quality, regardless of performance" (this can be used for example by napari-animation, or to control *which* layer the user really wants to keep high quality while sacrificing others)

Reply:
They don't have the same value, they are told one of increase, decrease, set to min, or set to max. How they do that is up to each layer, and it's relative to the present state.

Reply:
But they are all told "decrease" at the same time. I just don't get the benefit of all this duplication and having extra state on the layer that's not layer-specific but render-specific. What is the benefit?

Reply:
Ok, so the idea is to have a smarter loop in the future that selectively reduces the quality of specific layers starting from the most "expensive" one?

I think you still misunderstandmy point: I don't want to go "VisPy/Qt only". I'm also saying to stay in model-land in terms of the  "increase/decrease" signal. I'm just proposing to have a single value up in the viewer that says "TOO SLOW" and the layers receive the message, rather than setting the exact same value on each layer saying "TOO SLOW".